Fix ODR problems and includes in Selections.h

Teemperor opened the pull request:
We include MessageLogger.h because we use edm::LogVerbatim in this
header and this way it can compile on its own.

We renamed Selection to FilterSelection because we already have
a Selection class in CommonTools/Utils/interface/Selection.h,
which is breaking the One Definition Rule and breaks the C++
modules builds (and can also break the normal CMSSW builds).

Teemperor commented:
Part of the work going on regarding the C++ modules migration of CMSSW (tracked as issue https://github.com/cms-sw/cmssw/issues/15248). This PR is not meant to refactor things, but just to make these headers compile.

cmsbuild commented:
A new Pull Request was created by  @Teemperor (Raphael Isemann) for master.

It involves the following packages:

PhysicsTools/UtilAlgos


@cmsbuild, @monttj, @davidlange6 can you please review it and eventually sign? Thanks.
@davidlange6 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 3d883057ddc84ef9b58fb026cf5c408b66c01d10

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19353/20753/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  ClangBuild**

* **Build**:

I found an error when building: 
 
 <pre>>> Compiling edm plugin /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_2_X_2017-06-19-1100/src/PhysicsTools/UtilAlgos/plugins/ConfigurableAnalysis.cc 
>> Compiling edm plugin /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_2_X_2017-06-19-1100/src/PhysicsTools/UtilAlgos/plugins/NTuplingDevice.cc 
>> Compiling edm plugin /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_2_X_2017-06-19-1100/src/PhysicsTools/UtilAlgos/plugins/PrimaryVertexFilter.cc 
In file included from /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_2_X_2017-06-19-1100/src/PhysicsTools/UtilAlgos/plugins/ConfigurableAnalysis.cc:36:0:
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_2_X_2017-06-19-1100/src/PhysicsTools/UtilAlgos/interface/Selections.h: In constructor 'FilterSelections::FilterSelections(const edm::ParameterSet&, edm::ConsumesCollector&&)':
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_2_X_2017-06-19-1100/src/PhysicsTools/UtilAlgos/interface/Selections.h:389:22: error: 'Selection' was not declared in this scope
     for (std::vector<Selection>::iterator sIt=selections_.begin();sIt!=selections_.end();++sIt){
                      ^
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_2_X_2017-06-19-1100/src/PhysicsTools/UtilAlgos/interface/Selections.h:389:31: error: template argument 1 is invalid
     for (std::vector<Selection>::iterator sIt=selections_.begin();sIt!=selections_.end();++sIt){
                               ^
</pre>

* **Clang**:

I found a compilation error while trying to compile with clang: 
I used this command: 
 scram b vclean && scram build -k -j 16 USER_CXXFLAGS='-fsyntax-only' COMPILER='llvm compile'


 <pre>>> Compiling  /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_2_X_2017-06-19-1100/src/PhysicsTools/UtilAlgos/src/BasicMuonAnalyzer.cc 
>> Compiling  /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_2_X_2017-06-19-1100/src/PhysicsTools/UtilAlgos/src/StringBasedNTupler.cc 
>> Compiling  /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_2_X_2017-06-19-1100/src/PhysicsTools/UtilAlgos/bin/FWLiteWithBasicAnalyzer.cc 
>> Compiling  /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_2_X_2017-06-19-1100/src/PhysicsTools/UtilAlgos/bin/mergeTFileServiceHistograms.cpp 
In file included from /build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_2_X_2017-06-19-1100/src/PhysicsTools/UtilAlgos/plugins/ConfigurableAnalysis.cc:36:
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_2_X_2017-06-19-1100/src/PhysicsTools/UtilAlgos/interface/Selections.h:389:22: error: use of undeclared identifier 'Selection'
    for (std::vector<Selection>::iterator sIt=selections_.begin();sIt!=selections_.end();++sIt){
                     ^
/build/cmsbld/jenkins-workarea/workspace/ib-any-integration/CMSSW_9_2_X_2017-06-19-1100/src/PhysicsTools/UtilAlgos/interface/Selections.h:389:32: error: no type named 'iterator' in the global namespace; did you mean simply 'iterator'?
    for (std::vector<Selection>::iterator sIt=selections_.begin();sIt!=selections_.end();++sIt){
                               ^~~~~~~~~~
</pre>

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

Teemperor committed:
Fix ODR problems and includes in Selections.h

We include MessageLogger.h because we use edm::LogVerbatim in this
header and this way it can compile on its own.

We renamed Selection to FilterSelection because we already have
a Selection class in CommonTools/Utils/interface/Selection.h,
which is breaking the One Definition Rule and breaks the C++
modules builds (and can also break the normal CMSSW builds).

cmsbuild commented:
Pull request #19353 was updated. @cmsbuild, @monttj, @davidlange6 can you please check and sign again.

Teemperor commented:
Fixed compilation errors.

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 52c8d85c7f65d4231c6a3a569d07e1d0a6e1636d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19353/20775/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-19353/20775/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 22
 - DQMHistoTests: Total histograms compared: 1803260
 - DQMHistoTests: Total failures: 44267
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 1758827
 - DQMHistoTests: Total skipped: 166
 - DQMHistoTests: Total Missing objects: 0
 - Checked 90 log files, 14 edm output root files, 22 DQM output files

davidlange6 commented:
merge
